### PR TITLE
feat(ecs): Basic data model and type checking

### PR DESCRIPTION
**Pull request by Ersikan**

### CODE REVIEW

File: libecs/include/pivot/ecs/Core/Data/type.hxx [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    /// The property is of type std::string
+    String,
+
+    /// The property is of type double
+    Number,
+
+    /// The property is of type int
+    Integer,
+
+    /// The property is of type boolean
```
Comment (m-bdf) [inline libecs/include/pivot/ecs/Core/Data/type.hxx]: bool

File: libecs/include/pivot/ecs/Core/Data/type.hxx [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    Vec3,
+};
+
+struct Type;
+
+/// The record type specifies the property names and types of a record
+using RecordType = std::map<std::string, Type>;
+
+/// A type in the pivot data model
+struct Type : public std::variant<BasicType, RecordType> {
```
Comment (m-bdf) [inline libecs/include/pivot/ecs/Core/Data/type.hxx:37]: no public

File: libecs/source/Core/Data/value.cxx [hunk 3]
```diff
@@ -0,0 +1,29 @@
+#include <pivot/ecs/Core/Data/value.hxx>
+
+namespace pivot::ecs::data
+{
+
+RecordType Record::type() const
+{
+    RecordType type;
+
```
Comment (m-bdf) [inline libecs/source/Core/Data/value.cxx:9]: rm line

File: libecs/source/Core/Data/value.cxx [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    for (const auto &[key, value]: *this) { type.insert({key, value.type()}); }
+
```
Comment (m-bdf) [inline libecs/source/Core/Data/value.cxx:11]: rm line

**Review by m-bdf — APPROVED** [review]

**Review by zcorniere — APPROVED** [review]

**Review by Najo-z — APPROVED** [review]
Maybe a single type to represent numbers, not integer AND double.

**Pull request merged**